components/director/internal/domain/application/service.go

```diff
@@ -260,6 +260,30 @@ func (s *service) CreateFromTemplate(ctx context.Context, in model.ApplicationRe
 	return s.genericCreate(ctx, in, creator)
 }
 
+func (s *service) CreateManyIfNotExistsWithEventualTemplate(ctx context.Context, applicationInputs []model.ApplicationRegisterInput, templateTypes []string) error {
+	appsToAdd, err := s.filterUniqueNonExistingApplications(ctx, applicationInputs)
+	if err != nil {
+		return errors.Wrap(err, "while filtering unique and non-existing applications")
+	}
+
+	for index, a := range appsToAdd {
+		template := templateTypes[index]
```

Comment:
I'm not a fan of this hidden strong coupling between `applicationInputs` and `templateTypes` and then relying on the common `index`.

I suggest grouping them in a map all the way back before passing them to the `CreateManyIfNotExistsWithEventualTemplate` function instead of as two separate arguments to the function.

Furthermore, this coupling will immediately brake if there are duplicate systems as the index will mismatch after the `filterUniqueNonExistingApplications` functions runs.